### PR TITLE
[vendor] add kubernetes code-generator

### PR DESCRIPTION
**Pull request by surajssd**
This commits adds code-generator in the vendor, this will help the user
to run `make codegen` without having to setup `k8s.io/code-generator`
in their `GOPATH`. Also save us from pulling this repo during e2e tests
run in CircleCI.

Fixes https://github.com/habitat-sh/habitat-operator/issues/330

### CODE REVIEW

File: pkg/apis/habitat/v1beta1/zz_generated.deepcopy.go [hunk 1]
```diff
@@ -48,12 +48,8 @@ func (in *Habitat) DeepCopyInto(out *Habitat) {
 	out.Status = in.Status
 	if in.CustomVersion != nil {
 		in, out := &in.CustomVersion, &out.CustomVersion
-		if *in == nil {
```
Comment (krnowak) [inline pkg/apis/habitat/v1beta1/zz_generated.deepcopy.go]: The change in generated files tell me that you have imported a different version of the code generator from what was used before to generate those files… Not sure what to do with this fact. I mean - you seem to have used 1.10.1 which looks alright.

I think I'll let it slide, it is just two files of the bazilions that had been changed by this update.

File: .circleci/config.yml [hunk 2]
```diff
@@ -20,19 +20,6 @@ defaults: &defaults
         name: setup kubectl
         command: |
           curl -Lo kubectl https://storage.googleapis.com/kubernetes-release/release/${K8S_VERSION}/bin/linux/amd64/kubectl && chmod +x kubectl && sudo mv kubectl /usr/local/bin/
-    - run:
```
Comment (krnowak) [inline .circleci/config.yml]: No no no. You should have dropped only the part where we create the `$CODEGEN_PKG` directory and do a `git clone` of the code generator repo. The `hack/verify-codegen.sh` needs to remain. This is where our CI checks if you have actually update the generated code after changing the CRD.

File: Makefile [hunk 3]
```diff
@@ -70,5 +70,5 @@ update-version:
 
 .PHONY: codegen
 codegen:
-	CODEGEN_PKG=../../../k8s.io/code-generator hack/update-codegen.sh
+	CODEGEN_PKG=./vendor/k8s.io/code-generator hack/update-codegen.sh
```
Comment (krnowak) [inline Makefile]: Could you maybe modify both `hack/update-codegen.sh` and `hack/verify-codegen.sh` to drop this `CODEGEN_PKG` altogether? The scripts could just do something like:
```
readonly codegen_dir="$(dirname "${BASH_SOURCE[0]}")/../vendor/k8s.io/code-generator"`
```

**Comment by krnowak** [comment]
The version of codegen we use is based on the latest supported version of kubernetes. We don't regenerate code for older ones.

File: pkg/apis/habitat/v1beta1/zz_generated.deepcopy.go (hunk 1 above)
Comment (surajssd) [reply pkg/apis/habitat/v1beta1/zz_generated.deepcopy.go]: sure :-)

File: .circleci/config.yml (hunk 2 above)
Comment (surajssd) [reply .circleci/config.yml]: done

File: Makefile (hunk 3 above)
Comment (surajssd) [reply Makefile]: done

**Comment by surajssd** [comment]
@indradhanush we have pin pointed the version that is compatible with the code we have in `Gopkg.toml`, so developer need not have to do any checkouts.

**Comment by indradhanush** [comment]
Thanks for clarifying! :+1:

File: .circleci/config.yml [hunk 4]
```diff
@@ -23,16 +23,9 @@ defaults: &defaults
     - run:
         name: code-gen script
         environment:
-          CODEGEN_PKG: ../../../../src/k8s.io/code-generator
           GOPATH: /home/circleci/.go_workspace
         command: |
-          if [[ -n $CODEGEN_VERSION ]]; then
```
Comment (krnowak) [inline .circleci/config.yml]: This is a nice simplification, thanks.

Could you search for `CODEGEN_VERSION` in this file and just drop it? I think it was only used here.

File: Gopkg.toml [hunk 5]
```diff
@@ -20,11 +20,33 @@
 #  name = "github.com/x/y"
 #  version = "2.4.0"
 
+# Force dep to vendor the code generators, which aren't imported just used at dev time.
+required = [
+  "k8s.io/code-generator/cmd/defaulter-gen",
```
Comment (krnowak) [inline Gopkg.toml]: Pity that there isn't just a single package that will pull all the generators. But hopefully, if a new generator is added in some future version of k8s and we forget to add it while vendoring a new version, the generator will fail, so we will know to update the list.

I wonder if we could do something that will make us look at that list every time we vendor a new version of k8s to verify if this list is up-to-date.

Comment (krnowak) [reply Gopkg.toml]: Alright, then I suppose it's fine.

Comment (surajssd) [reply Gopkg.toml]: is it good to go?

Comment (krnowak) [reply Gopkg.toml]: Previous comment was not yet addressed - https://github.com/habitat-sh/habitat-operator/pull/337#discussion_r211247256

File: .circleci/config.yml (hunk 4 above)
Comment (surajssd) [reply .circleci/config.yml]: done :-)

**Pull request merged**